workflow/controller/dag.go

```diff
@@ -624,6 +624,13 @@ func (d *dagContext) evaluateDependsLogic(taskName string) (bool, bool, error) {
 			return false, false, nil
 		}
 
+		// If a task happens to have an onExit node, don't proceed until the onExit node is fulfilled
+		if onExitNode := d.wf.GetNodeByName(common.GenerateOnExitNodeName(taskName)); onExitNode != nil {
+			if !onExitNode.Fulfilled() {
+				return false, false, nil
+			}
+		}
+
```

Comment:
Note that this is not a regression of 2.9 or the DAG refactor, rather an already existing bug that hasn't been brought up before: in DAGs, the `onExit` handler of a task runs concurrently with the task's dependent steps, rather than waiting for the `onExit` handler to finish before running, as is done in steps